### PR TITLE
[JENKINS-20546] Preserve symlinks when copying artifacts

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-20546](https://issues.jenkins-ci.org/browse/JENKINS-20546)

### CODE REVIEW

**Comment by jglick** [comment]
@reviewbybees

**Comment by jglick** [comment]
Note: could be backported to 1.33.1-SNAPSHOT but the test fails in the setup because `FilePath.copyRecursiveTo` did not handle symlinks in 1.509 and throws an exception while archiving from the upstream project. Updating the parent to 1.532.3 avoids the exception, but the symlinks are skipped so the test still fails later because they are just absent from the downstream project. Updating the parent to 1.554.3 (or 1.554) lets the test pass.

**Comment by recampbell** [comment]
:+1:

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java [hunk 1]
```diff
@@ -1662,6 +1662,28 @@ public void testFilePermission() throws Exception {
             assertEquals(0755, w.child("artifactWithExecuteInSubdir.txt").mode() & 0777);
         }
     }
+
+    @Bug(20546)
```
Comment (tfennelly) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: Is `Bug` not deprecated?

**Comment by tfennelly** [comment]
:+1:

File: src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java (hunk 1 above)
Comment (jglick) [inline src/test/java/hudson/plugins/copyartifact/CopyArtifactTest.java]: Yes but using it here makes backporting slightly simpler.

**Comment by ikedam** [comment]
+1

I don't think we need to backport this to 1.33 as this isn't a critical issue and there are workarounds (like packing files with tar command).
Backporting is required only for security fixes.

**Comment by ikedam** [comment]
Notes
* On Windows, tests need administrative privilege to pass. Windows requires administrative privilege to create symbolic links.
* The test requires Java >= 7. `FilePath#readLink` works only with Java >= 7.
    * I always use Java6 to make a release as I faced a problem that occurred only when compiled with Java6 ([JENKINS-21274](https://issues.jenkins-ci.org/browse/JENKINS-21274)). I used Java7 to release copyartifact 1.35.
    * (That should work even compiled with Java6 only if running with Java>=7)

**Pull request merged**